Strict whitelisting

tonylepmets opened the pull request:
Change whitelisting approach. For masking do not include child nodes if
the parent node itself is not whitelisted. Added prune action which only
includes a field if it is whitelisted. Ported from ruby logasm project.

take-five commented:
it's a public repository, I don't think we should mention tickets in commit messages

take-five commented:
Also, for ruby we had to go further and exclude non-whitelisted fields from the output completely, because services generated a lot of unique fields in non-root attributes. So, if "foo/bar" attribute was whitelisted, all sub-keys of "foo" were included in logs (masked).

tonylepmets commented:
Sounds reasonable, I'll add the option to specify the `prune` strategy.

tonylepmets committed:
Apply strict whitelisting

Change whitelisting approach. For masking do not include child nodes if
the parent node itself is not whitelisted. Added prune action which only
includes a field if it is whitelisted. Ported from ruby logasm project.

tonylepmets committed:
Remove coffee compiler flag from mocha opts

Project moved to JS

tonylepmets committed:
Bump version to 1.3.0

tonylepmets commented:
@take-five updated. Ported the ruby logasm solution.

indrekj commented on lib/logasm/preprocessors/strategies/mask_strategy.js:
```diff
@@ -0,0 +1,59 @@
+const WILDCARD_SYMBOL = '~';
+const MASKED_VALUE = '*****';
+
+class MaskStrategy {
```
Btw, you can also encapsulate objects like
```js
function createMaskStrategy(whitelistedFields) {
  const somePrivateFunc = ...;
  const process = function(data, path) {...}

  return {process: process};
}
module.exports = createMaskStrategy;
```
We used this style in Medra a lot. This makes `this.` redundant so you don't have to worry about scoping at all. You also don't need `_functions` for private functions which are not really private. We found it quite useful.

Not forcing this change though.

indrekj commented on lib/logasm/preprocessors/strategies/mask_strategy.js:
```diff
@@ -0,0 +1,59 @@
+    }
+  }
+
+  _isWhitelistedField(path) {
```
should we cache the result? Btw, do you have the performance results compared to ruby? Even though the performance did not degrade I assume, it would be cool to see how both of the libraries compare to each other

tonylepmets replied:
Could cache for better performance yeah. What do you think of using this https://github.com/groupon/node-cached Has a similar getset function as we're using in ruby logasm.

indrekj replied:
I guess it can stay for now
